python/ray/job_config.py

```diff
@@ -44,4 +43,7 @@ def serialize(self):
             self.num_java_workers_per_process)
         job_config.jvm_options.extend(self.jvm_options)
         job_config.code_search_path.extend(self.code_search_path)
+        if self.runtime_env.get("working_dir_uri"):
```

Comment:
Why isn't this `job_config.runtime_env = copy.deepcopy(self.runtime_env)`?

Reply:
`job_config = ray.gcs_utils.JobConfig()` is defined in protobuf and it's not what exposed to end-users.
`self ` is a python class. 

In protobuf, RuntimeEnv is a structure. I think it's better to make it a structure there since the data will be accessed in GCS and typed data makes it easier to do it.

Expose a weak-typed dict to end-user is reasonable, but for internal used structure, strong typed structure is better. Besides, internally, we don't need anything about `working_dir` or `modules`, only `uri`s are required.

We actually expose a user-facing python-based `json_config` to users, but a protobuf based `job_config` in all raylet.

Reply:
Can we make a utility function to convert from JSON to protobuf then? Then we can write

job_config.runtime_env = runtime_env_json_to_proto(self.runtime_env)

which is clearer to read.

Reply:
That makes sense.